src/lib/pubkey/ec_group/ec_group.cpp

```diff
@@ -72,97 +65,106 @@ class EC_Group_Data_Map final {
+            if(i->oid() == oid) {
+               /*
+               * If both OID and params are the same then we are done, just return
+               * the already registered curve obj.
+               *
+               * If the params don't match then the user attempted to create some
+               * other set of parameters under an already known OID
+               */
+               if(i->params_match(p, a, b, g_x, g_y, order, cofactor)) {
+                  return i;
+               } else {
+                  throw Invalid_Argument("Attempting to register a curve using OID " + oid.to_string() +
+                                         " but a distinct curve is already registered using that OID");
+               }
             }
 
             /*
             * If the same curve was previously created without an OID but is now
             * being registered again using an OID, save that OID.
+            *
+            * TODO(Botan4) remove this block; this situation won't be possible since
+            * we will require all groups to have an OID
             */
-            if(same_params && i->oid().empty() && !oid.empty()) {
+            if(i->oid().empty() && i->params_match(p, a, b, g_x, g_y, order, cofactor)) {
                i->set_oid(oid);
                return i;
             }
          }
 
          /*
          Not found in current list, so we need to create a new entry
-
-         If an OID is set, try to look up relative our static tables to detect a duplicate
-         registration under an OID
          */
-
-         auto new_group = EC_Group_Data::create(p, a, b, g_x, g_y, order, cofactor, oid, source);
-
-         if(oid.has_value()) {
-            std::shared_ptr<EC_Group_Data> data = EC_Group::EC_group_info(oid);
-            if(data != nullptr && !new_group->params_match(*data)) {
+         auto new_group = EC_Group::EC_group_info(oid);
+         if(new_group) {
+            /*
+            * This turned out to be the OID of one of the builtin groups. Verify
+            * that all of the provided parameters match that builtin group.
+            */
+            if(!new_group->params_match(p, a, b, g_x, g_y, order, cofactor)) {
                throw Invalid_Argument("Attempting to register an EC group under OID of hardcoded group");
             }
          } else {
-            // Here try to use the order as a hint to look up the group id, to identify common groups
-            const OID oid_from_store = EC_Group::EC_group_identity_from_order(order);
-            if(oid_from_store.has_value()) {
-               std::shared_ptr<EC_Group_Data> data = EC_Group::EC_group_info(oid_from_store);
+            /*
+            * This path is taken for an application registering a new EC_Group with an OID specified
+            */
+            new_group = EC_Group_Data::create(p, a, b, g_x, g_y, order, cofactor, oid, source);
+         }
```

Comment:
Another nit (trying to avoid the mutable variable):

```C++
         auto new_group = [&] {
            if(auto g = EC_Group::EC_group_info(oid); g != nullptr) {
               /*
               * This turned out to be the OID of one of the builtin groups. Verify
               * that all of the provided parameters match that builtin group.
               */
               if(!g->params_match(p, a, b, g_x, g_y, order, cofactor)) {
                  throw Invalid_Argument("Attempting to register an EC group under OID of hardcoded group");
               }

               return g;
            }
            
            return EC_Group_Data::create(p, a, b, g_x, g_y, order, cofactor, oid, source);
         }();
```